src/Common/Threading.h

```diff
@@ -0,0 +1,270 @@
+        void remove(CancelToken * token);
+        void signal(UInt64 tid);
+        void signal(UInt64 tid, int code, const String & message);
+
+        static const std::shared_ptr<Registry> & instance();
+    };
+
+    // Cancels this token and wakes thread if necessary.
+    // Can be called from any thread.
+    void signalImpl();
+    void signalImpl(int code, const String & message);
+
+    // Lower bit: cancel signal received flag
+    static constexpr UInt64 canceled = 1;
+
+    // Upper bits - possible values:
+    // 1) all zeros: token is enabed, i.e. wait() call can return false, thread is not waiting on any address;
+    // 2) all ones: token is disabled, i.e. wait() call cannot be cancelled;
+    // 3) specific `address`: token is enabled and thread is currently waiting on this `address`.
+    static constexpr UInt64 disabled = ~canceled;
+    static_assert(sizeof(UInt32 *) == sizeof(UInt64)); // State must be able to hold an address
+
+    // All signal handling logic should be globally serialized using this mutex
+    static std::mutex signal_mutex;
+
+    // Cancellation state
+    alignas(64) std::atomic<UInt64> state;
+    [[maybe_unused]] char padding[64 - sizeof(state)];
+
+    // Cancellation exception
+    int exception_code;
+    String exception_message;
+
+    // Token is permanently attached to a single thread. There is one-to-one mapping between threads and tokens.
+    const UInt64 thread_id;
+
+    // To avoid `Registry` destruction before last `Token` destruction
+    const std::shared_ptr<Registry> registry;
+};
+
+class CancellableSharedMutex
+{
+public:
+    CancellableSharedMutex();
+    ~CancellableSharedMutex() = default;
+    CancellableSharedMutex(const CancellableSharedMutex &) = delete;
+    CancellableSharedMutex & operator=(const CancellableSharedMutex &) = delete;
+
+    // Exclusive ownership
+    void lock();
+    bool try_lock();
+    void unlock();
+
+    // Shared ownership
+    void lock_shared();
+    bool try_lock_shared();
+    void unlock_shared();
+
+private:
+    // State 64-bits layout:
+    //    1b    -   31b   -    1b    -   31b
+    // signaled - writers - signaled - readers
+    // 63------------------------------------0
+    // Two 32-bit words are used for cancellable waiting, so each has its own separate signaled bit
+    static constexpr UInt64 readers = (1ull << 32ull) - 1ull - CancelToken::signaled;
+    static constexpr UInt64 readers_signaled = CancelToken::signaled;
+    static constexpr UInt64 writers = readers << 32ull;
+    static constexpr UInt64 writers_signaled = readers_signaled << 32ull;
+
+    alignas(64) std::atomic<UInt64> state;
+    std::atomic<UInt32> waiters;
+};
+
+class FastSharedMutex
+{
+public:
+    FastSharedMutex();
+    ~FastSharedMutex() = default;
+    FastSharedMutex(const FastSharedMutex &) = delete;
+    FastSharedMutex & operator=(const FastSharedMutex &) = delete;
+
+    // Exclusive ownership
+    void lock();
+    bool try_lock();
```

Comment:
Here accessing `state` on write will invalidate `waiters` and vise versa. I don't think it's desirable. The same is above in `CancelableSharedMutex`

Reply:
I'm not sure I totally understand your concern. But actually, `waiters` is an optimization, to avoid making syscall in vain (e.g. when we unlock and nobody waits). So it can be "out of sync" with `state` unless it violates correctness and now I cannot see how it can be violated.

Reply:
Ah, I got what you are talking about. They are in the same cache line. Yes, I benchmark it and having them in the same cache line turned out to be faster. Maybe just because there is less cache lines to access/sync. Not sure why, but for tests we currently have it's better.